edb/edgeql/compiler/inference/multiplicity.py

```diff
@@ -0,0 +1,701 @@
+            return result
+
+        # Even when arguments are of multiplicity ONE, we cannot
+        # exclude the possibility of the result being of multiplicity
+        # MANY. We need to check that at most one argument has
+        # cardinality more than ONE.
+
+        if len([card for card in cards if card.is_multi()]) > 1:
+            return MANY
+        else:
+            return result
+
+
+@_infer_multiplicity.register
+def __infer_const(
+    ir: irast.BaseConstant,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return ONE
+
+
+@_infer_multiplicity.register
+def __infer_param(
+    ir: irast.Parameter,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return ONE
+
+
+@_infer_multiplicity.register
+def __infer_const_set(
+    ir: irast.ConstantSet,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    if len(ir.elements) == len({el.value for el in ir.elements}):
+        return ONE
+    else:
+        return MANY
+
+
+@_infer_multiplicity.register
+def __infer_typecheckop(
+    ir: irast.TypeCheckOp,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    # Unless this is a singleton, the multiplicity cannot be assumed to be ONE.
+    card = cardinality.infer_cardinality(
+        ir, scope_tree=scope_tree, ctx=ctx)
+    if card is not None and card.is_single():
+        return ONE
+    else:
+        return MANY
+
+
+@_infer_multiplicity.register
+def __infer_typecast(
+    ir: irast.TypeCast,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    return infer_multiplicity(
+        ir.expr, scope_tree=scope_tree, ctx=ctx,
+    )
+
+
+def _infer_stmt_multiplicity(
+    ir: irast.FilteredStmt,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+    result = infer_multiplicity(
+        ir.subject if isinstance(ir, irast.MutatingStmt) else ir.result,
+        scope_tree=scope_tree,
+        ctx=ctx,
+    )
+
+    # WITH block bindings need to be validated, they don't have to
+    # have multiplicity ONE, but their sub-expressions must be valid.
+    #
+    # Inferring how the FILTER clause affects multiplicity is in
+    # general impossible, but we still want to ensure that the FILTER
+    # expression has valid multiplicity.
+    for part in ir.bindings + [ir.where]:
+        if part:
+            infer_multiplicity(part, scope_tree=scope_tree, ctx=ctx)
+
+    return result
+
+
+def _infer_for_multiplicity(
+    ir: irast.SelectStmt,
+    *,
+    scope_tree: irast.ScopeTreeNode,
+    ctx: inference_context.InfCtx,
+) -> qltypes.Multiplicity:
+
+    assert ir.iterator_stmt is not None
+    itexpr = ir.iterator_stmt.expr
+
+    if not isinstance(ir.result, irast.Set):
```

Comment:
`SelectStmt.result` is always a `Set`, this check is redundant.

Reply:
Removed.